source/common/tcp_proxy/upstream.cc

```diff
@@ -244,31 +246,55 @@ void Http2Upstream::setRequestEncoder(Http::RequestEncoder& request_encoder, boo
   request_encoder_->getStream().addCallbacks(*this);
   const std::string& scheme =
       is_ssl ? Http::Headers::get().SchemeValues.Https : Http::Headers::get().SchemeValues.Http;
-  auto headers = Http::createHeaderMap<Http::RequestHeaderMapImpl>(
-      {{Http::Headers::get().Method, "CONNECT"},
-       {Http::Headers::get().Protocol, Http::Headers::get().ProtocolValues.Bytestream},
-       {Http::Headers::get().Scheme, scheme},
-       {Http::Headers::get().Path, "/"},
-       {Http::Headers::get().Host, hostname_}});
+  auto headers = Http::createHeaderMap<Http::RequestHeaderMapImpl>({
+      {Http::Headers::get().Method, config_.use_post() ? "POST" : "CONNECT"},
+      {Http::Headers::get().Host, config_.hostname()},
+      {Http::Headers::get().Path, "/"},
+      {Http::Headers::get().Scheme, scheme},
+  });
+
+  if (!config_.use_post()) {
+    headers->addReference(Http::Headers::get().Protocol,
+                          Http::Headers::get().ProtocolValues.Bytestream);
+  }
+
+  for (const auto& header : config_.headers()) {
+    if (header.append().value()) {
+      headers->addCopy(Http::LowerCaseString(header.header().key()), header.header().value());
```

Comment:
definitely being able to append to :-prefixed headers seems buggy

Reply:
could you help point me to example to exclude : prefixed headers and host header?

Reply:
It doesn't seem that other places in the code base that also add addition request headers have any check though ..

Reply:
https://github.com/envoyproxy/envoy/blob/master/source/common/router/header_parser.cc#L51

They do it on config ingress, not application, and I think we should do the same here.

Reply:
thx!